src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs

```diff
@@ -491,6 +542,124 @@ private int FindEndPosition(ReadOnlySpan<char> input, int i)
+            }
+
+            // Use two maps from state IDs to register values for the current and next set of states.
+            // Note that these maps use insertion order, which is used to maintain priorities between states in a way
+            // that matches the order the backtracking engines visit paths.
+            SparseIntMap<Registers> current = new(), next = new();
+            current.Add(state.Id, initialRegisters);
+
+            while (i < input.Length)
+            {
+                Debug.Assert(next.Count == 0);
+
+                TSetType[]? minterms = _builder._minterms;
+                Debug.Assert(minterms is not null);
+
+                int c = input[i];
+                int normalMintermId = _partitions.GetMintermID(c);
+
+                foreach (var (sourceId, sourceRegisters) in current.Values)
+                {
+                    Debug.Assert(_builder._statearray is not null);
+                    DfaMatchingState<TSetType> sourceState = _builder._statearray[sourceId];
+
+                    // Find the minterm, handling the special case for the last \n
+                    int mintermId = c == '\n' && i == input.Length - 1 && sourceState.StartsWithLineAnchor ?
+                        minterms.Length : normalMintermId; // mintermId = minterms.Length represents \Z (last \n)
+                    TSetType minterm = (uint)mintermId < (uint)minterms.Length ?
+                        minterms[mintermId] :
+                        _builder._solver.False; // minterm=False represents \Z
+
+                    // Get or create the transitions
+                    int offset = (sourceId << _builder._mintermsCount) | mintermId;
+                    Debug.Assert(_builder._capturingDelta is not null);
+                    var transitions = Volatile.Read(ref _builder._capturingDelta[offset])
+                        ?? CreateNewCapturingTransitions(sourceState, minterm, offset);
+
+                    // Take the transitions in their prioritized order
+                    for (int j = 0; j < transitions.Count; ++j)
+                    {
+                        var (targetState, effects) = transitions[j];
+                        if (targetState.IsDeadend)
+                            continue;
+                        // Try to add the state and handle the case where it didn't exist before. If the state already
+                        // exists, then the transition can be safely ignored, as the existing state was generated by a
+                        // higher priority transition.
+                        if (next.Add(targetState.Id, out int index))
+                        {
+                            // Avoid copying the registers on the last transition from this state, reusing the registers instead
+                            Registers newRegisters = j != transitions.Count - 1 ? sourceRegisters.Clone() : sourceRegisters;
```

Comment:
Am I interpreting this correctly that we're allocating two new arrays of the length of the number of captures for every step for every transition, so effectively O(MatchLength * AvgStatesPerNode * NumCaptures * 2)?

Reply:
Doesn't seem like it would be too hard to add an optional memoization capability to RegexOptions.Compiled, such that we cache whether we've already examined a previous input location for a given node in the pattern.  This is one of the ideas we'd bandied about in https://github.com/dotnet/runtime/issues/1349, but here we could choose to have NonBacktracking create a RegexOptions.Compiled instance with that feature enabled, for the purposes of this third leg.  Would that be valuable?

Reply:
I think it could be at least situationally. Based on the Rust regex comments they use it for short matches to keep the size of the bitmap reasonable. So, patterns that generate a dense set of short matches could especially benefit.